test/controllers/orders_controller_test.rb

```diff
@@ -0,0 +1,274 @@
+require "test_helper"
+
+describe OrdersController do
+
+  describe "index" do
+    it "sends success when the user is signed in" do
+      login(User.first)
+
+      get orders_path
+
+      must_respond_with :success
+    end
+
+    it "responds with unauthorized if not signed in" do
+      get orders_path
+
+      must_respond_with :unauthorized
+    end
+
+  end
+
+  describe "show" do
+    it "responds with success when cart exists" do
+      get cart_path
+
+      get order_path(Order.find_by(id: session[:order_id]))
+
+      must_respond_with :success
+    end
+
+    it "responds with success when cart doesnt exist" do
+      order = Order.create!
+      Cartitem.create(order: order, product: Product.first, quantity: 3)
+
+      get order_path(order)
+
+      must_respond_with :success
+    end
+
+  end
+
+  describe "edit" do
+    it "responds with success if the order exists" do
+      get order_path(Order.first)
+
+      must_respond_with :success
+    end
+
+    it "responds with not found if the order doesn't exists" do
+      order_id = Order.last.id + 1
+      get order_path(order_id)
+
+      must_respond_with :not_found
+    end
+  end
+
+
+  describe "update" do
+
+    before do
+      @order = Order.first
+      @order_data = {
+        name: "test customer",
+        email: "something",
+        ccnumber: "12345678901234567",
+        ccexpiration: Date.today + 60.days,
+        cvv: "234",
+        address: "some test address",
+        zipcode: "90235"
+      }
+    end
+
+
+
+    it "updates an existing order with valid data" do
+
+      old_order_count = Order.count
+
+      patch order_path(@order), params: { order: @order_data }
+
+      must_respond_with :success
+      @order.reload
+      @order.name.must_equal @order_data[:name]
+      Order.count.must_equal old_order_count
+    end
+
+
+
+    it "won't update existing order with invalid data" do
+
+      # with past date
+      @order_data[:ccexpiration] = Date.today - 2.months
+
+      old_order_count = Order.count
+
+      patch order_path(@order), params: { order: @order_data }
+
+      must_respond_with :bad_request
+      @order.reload
+      @order.name.wont_equal @order_data[:name]
+      Order.count.must_equal old_order_count
+
+
+      # with future date more than 5 years
```

Comment:
These are all great test cases, and I'd recommend that they'd all be their own test. When trying to figure out if things should be their own test, I try to think: Is there more than one specific set of code that would need to change to fail this test? If so, it might be good to split it out.